File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/core/BoxTrustStatusReport.kt
// Kotlin mirror of packages/protocol/src/boxTrustStatus.ts — the STK-signed
// PER-BOX relay-trust verdict.
//
// Each box independently verifies the ServiceBlessing it is handed by the
// `.services` hub; that verdict is genuinely per-box. The daemon signs this
// tuple with its STK (Ed25519); `.com` relays it VERBATIM on `/pods`
// (`trustStatus: { report, signatureHex }`). A phone that derives the box STK
// locally (ServerKeys.deriveStkPub — NOT `.com`'s echo) re-verifies it
// end-to-end, so a rogue `.com` can DROP a report but cannot FORGE one: the
// per-server warning a client renders is the box's own word.
//
// SIBLING of DaemonStatusReport — do NOT fold together. Canonical bytes MUST
// match the TS implementation byte-for-byte (pinned cross-platform vector in
// packages/protocol/tests/boxTrustStatus.test.ts + BoxTrustStatusReportTest):
//
//   flagship/box-trust-status/v1|<serverDomain>|<relayVerdict>|
//   <lockedDown "1"|"0">|<failingCertHash or "">|
//   <coveringExceptionCertHash or "">|<nonce>|<issuedAt>

package com.flagshipserver.app.core

import com.google.crypto.tink.subtle.Ed25519Verify

object BoxTrustStatusReport {
    const val CANONICAL_TAG = "flagship/box-trust-status/v1"

    /** Freshness bound on a relayed report, mirroring DaemonStatusReport. */
    const val MAX_REPORT_AGE_MS: Long = 7L * 24 * 60 * 60 * 1000

    /** The box's verdict on the relay-class (`.services` hub) blessing. */
    enum class RelayVerdict(val wire: String) {
        TRUSTED("trusted"),
        UNTRUSTED("untrusted"),
        UNKNOWN("unknown"),
        ;

        companion object {
            fun fromWire(s: String): RelayVerdict? = entries.firstOrNull { it.wire == s }
        }
    }

    data class Report(
        val serverDomain: String,
        val relayVerdict: RelayVerdict,
        val lockedDown: Boolean,
        /** relay-class cert-hash of the offending hub key, when untrusted. */
        val failingCertHash: String?,
        /** relay-class cert-hash of the owner TrustException that lifted the
         *  failing verdict, when an override is in force. */
        val coveringExceptionCertHash: String?,
        val nonce: String,
        val issuedAt: Long,
    )

    fun canonicalBytes(r: Report): ByteArray = listOf(
        CANONICAL_TAG,
        r.serverDomain,
        r.relayVerdict.wire,
        if (r.lockedDown) "1" else "0",
        r.failingCertHash ?: "",
        r.coveringExceptionCertHash ?: "",
        r.nonce,
        r.issuedAt.toString(),
    ).joinToString("|").toByteArray(Charsets.UTF_8)

    /** Verify the box's signature under its (locally derived) STK pubkey.
     *  Returns false — never throws — on malformed input. */
    fun verify(r: Report, signature: ByteArray, stkPub: ByteArray): Boolean = try {
        Ed25519Verify(stkPub).verify(signature, canonicalBytes(r))
        true
    } catch (_: Throwable) {
        false
    }
}

File: apps/mobile/shared/Sources/FlagshipAPI/BoxTrustStatusReport.swift
import Foundation

/// The box's verdict on the relay-class (`.services` hub) blessing it holds.
public enum RelayVerdict: String, Codable, Sendable {
    case trusted
    case untrusted
    case unknown
}

/// PER-BOX relay-trust verdict — the DATA half of the box's STK-signed
/// `flagship/box-trust-status/v1` envelope (canonical bytes + verify live in
/// `FlagshipCore.BoxTrustStatus`, mirroring the `DaemonStatusReport`/
/// `DaemonStatus` split so the wire type can ride `PodDirectoryEntry` without a
/// FlagshipCore ↔ FlagshipAPI cycle).
///
/// Each box independently verifies the ServiceBlessing the `.services` hub
/// hands it; that verdict is genuinely per-box. `.com` relays this VERBATIM on
/// `/pods` (`trustStatus: { report, signatureHex }`). A phone re-verifies it
/// under the box's registered STK, so a rogue `.com` can DROP a report but not
/// FORGE one. Clients aggregate warnings by `failingCertHash` ACROSS all a
/// user's pods: one sliver line + one override per DISTINCT faulty authority.
public struct BoxTrustStatusReport: Codable, Equatable, Sendable {
    public let serverDomain: String
    public let relayVerdict: RelayVerdict
    public let lockedDown: Bool
    /// relay-class cert-hash of the offending hub key, when untrusted.
    public let failingCertHash: String?
    /// relay-class cert-hash of the owner TrustException that lifted the
    /// failing verdict, when an override is in force.
    public let coveringExceptionCertHash: String?
    public let nonce: String
    public let issuedAt: Int64

    public init(
        serverDomain: String,
        relayVerdict: RelayVerdict,
        lockedDown: Bool,
        failingCertHash: String?,
        coveringExceptionCertHash: String?,
        nonce: String,
        issuedAt: Int64
    ) {
        self.serverDomain = serverDomain
        self.relayVerdict = relayVerdict
        self.lockedDown = lockedDown
        self.failingCertHash = failingCertHash
        self.coveringExceptionCertHash = coveringExceptionCertHash
        self.nonce = nonce
        self.issuedAt = issuedAt
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.serverDomain = try c.decode(String.self, forKey: .serverDomain)
        self.relayVerdict = try c.decode(RelayVerdict.self, forKey: .relayVerdict)
        self.lockedDown = try c.decodeIfPresent(Bool.self, forKey: .lockedDown) ?? false
        self.failingCertHash = try c.decodeIfPresent(String.self, forKey: .failingCertHash)
        self.coveringExceptionCertHash = try c.decodeIfPresent(String.self, forKey: .coveringExceptionCertHash)
        self.nonce = try c.decode(String.self, forKey: .nonce)
        self.issuedAt = try c.decode(Int64.self, forKey: .issuedAt)
    }

    private enum CodingKeys: String, CodingKey {
        case serverDomain, relayVerdict, lockedDown, failingCertHash, coveringExceptionCertHash, nonce, issuedAt
    }
}

/// `trustStatus` on a `/pods` pod: the verbatim box-trust-status report + its
/// Ed25519 signature (hex) under the box STK. Verified by
/// `FlagshipCore.BoxTrustStatus.verify`.
public struct SignedBoxTrustStatus: Codable, Equatable, Sendable {
    public let report: BoxTrustStatusReport
    public let signatureHex: String
    public init(report: BoxTrustStatusReport, signatureHex: String) {
        self.report = report
        self.signatureHex = signatureHex
    }
}
